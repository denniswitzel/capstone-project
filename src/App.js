import React, { useEffect, useState } from 'react'
import { useCookies } from 'react-cookie'
import { Route, Switch } from 'react-router-dom'
import styled from 'styled-components/macro'
import useFetch from '../src/hooks/useFetch'
import Categories from './components/Categories/Categories'
import ProductDetail from './components/ProductDetail/ProductDetail'
import ProductList from './components/ProductList/ProductList'
import { ReactComponent as Logo } from './images/icons/planternate-logo.svg'
import { categories } from './services/categories'
import { loadLocally, saveLocally } from './services/localStorage'
import RatingList from './components/RatingList/RatingList'

export default function App() {
  const [cookies, setCookies] = useCookies(['user'])

  const burger = useFetch('burger')
  const coldCuts = useFetch('cold-cuts')
  const kebab = useFetch('kebab')
  const minced = useFetch('minced')
  const nuggets = useFetch('nuggets')
  const sausages = useFetch('sausages')

  const [products, setProducts] = useState(loadLocally('products') ?? [])

  useEffect(() => {
    setProducts(burger.concat(coldCuts, kebab, minced, nuggets, sausages))
  }, [burger, coldCuts, kebab, minced, nuggets, sausages])

  useEffect(() => {
    saveLocally('products', products)
  }, [products])

  useEffect(() => {
    cookies.user ??
      setCookies('user', Math.floor(Math.random() * Math.floor(5000)), {
        path: '/',
      })
    // eslint-disable-next-line
  }, [])

  return (
    <Switch>
      <Route exact path="/">
        <LogoStyled />
        {categories?.map(({ name, icon }, index) => (
          <Categories key={index} categoryIcon={icon} categoryName={name} />
        ))}
      </Route>
      {categories.map(({ name, path }) => (
        <Route exact path={path} key={name}>
          <ProductList
            headline={name.toLocaleUpperCase()}
            product={products.filter((p) => p.category === name)}
          />
        </Route>
      ))}
      {categories.map(({ name, path }) => (
        <Route path={`${path}/:id`} key={name}>
          <ProductDetail
            product={products.filter((p) => p.category === name)}
            onFavoriteClick={toggleFavorite}
            cookies={cookies.user}
          />
        </Route>
      ))}
      <Route path="/favorites">
        <ProductList
          headline={'Favorites'}
          product={products
            .filter((p) => p.isFavorite)
            .sort((a, b) => a.title.localeCompare(b.title))}
        />
      </Route>
      <Route path="/rating">
        <RatingList product={products} />
      </Route>
    </Switch>
  )

  function toggleFavorite(favorite) {
    const index = products.findIndex(
      (favoriteItem) => favoriteItem.id === favorite.id
    )
    setProducts([
      ...products.slice(0, index),
      { ...favorite, isFavorite: !favorite.isFavorite },
      ...products.slice(index + 1),
    ])
  }
}

const LogoStyled = styled(Logo)`
  width: 60vw;
  grid-column: 1/3;
  margin: 30px 0 30px 0;
`
