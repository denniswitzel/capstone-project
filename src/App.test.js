import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import { render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import '@testing-library/jest-dom/extend-expect'
import App from './App'

describe('App', () => {
  it('has links to all pages', () => {
    const urls = [
      '/',
      '/burger',
      'burger/:id',
      '/kebab',
      '/kebab/:id',
      '/cold-cuts',
      '/sausages',
      '/sausages/:id',
      '/nuggets',
      '/nuggets/:id',
      '/minced',
      '/minced/:id',
    ]
    const { getAllByRole } = render(
      <Router>
        <App />
      </Router>
    )

    getAllByRole('link').forEach((link) => {
      expect(urls).toContain(link.getAttribute('href'))
    })
  })

  it('renders correctly', () => {
    const tree = renderer.create(
      <Router>
        <App />
      </Router>
    )
    expect(tree).toMatchSnapshot()
  })
})
