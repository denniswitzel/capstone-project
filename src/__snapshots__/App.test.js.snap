// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
Array [
  <svg
    className="App__LogoStyled-sc-3vbase-0 kqhSlf"
  >
    planternate-logo.svg
  </svg>,
  <a
    aria-current={null}
    className="sc-AxjAm eNUlSU"
    href="/burger"
    onClick={[Function]}
  >
    <div
      className="sc-AxirZ kGlEoH"
    >
      <img
        alt="burger"
        className="sc-AxiKw iMhkxe"
        src="https://raw.githubusercontent.com/denniswitzel/capstone-project/f892d3d147b3c573a796dcb1a34298351806d8b5/src/images/icons/burger.svg"
      />
      <h2
        className="sc-AxhCb liCkXb"
      >
        BURGER
      </h2>
    </div>
  </a>,
  <a
    aria-current={null}
    className="sc-AxjAm eNUlSU"
    href="/cold-cuts"
    onClick={[Function]}
  >
    <div
      className="sc-AxirZ kGlEoH"
    >
      <img
        alt="cold-cuts"
        className="sc-AxiKw iMhkxe"
        src="https://raw.githubusercontent.com/denniswitzel/capstone-project/717acadfd943ecc6bfc7f09650784cf00ff22b44/src/images/icons/coldcuts.svg"
      />
      <h2
        className="sc-AxhCb liCkXb"
      >
        COLD-CUTS
      </h2>
    </div>
  </a>,
  <a
    aria-current={null}
    className="sc-AxjAm eNUlSU"
    href="/kebab"
    onClick={[Function]}
  >
    <div
      className="sc-AxirZ kGlEoH"
    >
      <img
        alt="kebab"
        className="sc-AxiKw iMhkxe"
        src="https://raw.githubusercontent.com/denniswitzel/capstone-project/f892d3d147b3c573a796dcb1a34298351806d8b5/src/images/icons/kebab.svg"
      />
      <h2
        className="sc-AxhCb liCkXb"
      >
        KEBAB
      </h2>
    </div>
  </a>,
  <a
    aria-current={null}
    className="sc-AxjAm eNUlSU"
    href="/minced"
    onClick={[Function]}
  >
    <div
      className="sc-AxirZ kGlEoH"
    >
      <img
        alt="minced"
        className="sc-AxiKw iMhkxe"
        src="https://raw.githubusercontent.com/denniswitzel/capstone-project/f892d3d147b3c573a796dcb1a34298351806d8b5/src/images/icons/minced.svg"
      />
      <h2
        className="sc-AxhCb liCkXb"
      >
        MINCED
      </h2>
    </div>
  </a>,
  <a
    aria-current={null}
    className="sc-AxjAm eNUlSU"
    href="/nuggets"
    onClick={[Function]}
  >
    <div
      className="sc-AxirZ kGlEoH"
    >
      <img
        alt="nuggets"
        className="sc-AxiKw iMhkxe"
        src="https://raw.githubusercontent.com/denniswitzel/capstone-project/f892d3d147b3c573a796dcb1a34298351806d8b5/src/images/icons/nuggets.svg"
      />
      <h2
        className="sc-AxhCb liCkXb"
      >
        NUGGETS
      </h2>
    </div>
  </a>,
  <a
    aria-current={null}
    className="sc-AxjAm eNUlSU"
    href="/sausages"
    onClick={[Function]}
  >
    <div
      className="sc-AxirZ kGlEoH"
    >
      <img
        alt="sausages"
        className="sc-AxiKw iMhkxe"
        src="https://raw.githubusercontent.com/denniswitzel/capstone-project/67e1d46d4c80dc548f311f8b10cb5d5c61960c60/src/images/icons/sausage.svg"
      />
      <h2
        className="sc-AxhCb liCkXb"
      >
        SAUSAGES
      </h2>
    </div>
  </a>,
]
`;
